File: python/src/seventhplace/manager.py
# SPDX-License-Identifier: Apache-2.0
"""ConfigManager — orchestrates the layered configuration load pipeline."""

from __future__ import annotations

from pathlib import Path
from typing import Any, TypeVar

import pydantic

from seventhplace.errors import ValidationError
from seventhplace.merge import deep_merge
from seventhplace.providers.cli import CLIProvider
from seventhplace.providers.defaults import DefaultProvider
from seventhplace.providers.env import EnvProvider
from seventhplace.providers.file import FileProvider

T = TypeVar("T", bound=pydantic.BaseModel)


class ConfigManager:
    """Orchestrates the layered configuration load pipeline.

    Load flow:
        1. Extract defaults from schema (model_dump)
        2. Deep merge file overrides (one or more files, in order)
        3. Deep merge env overrides
        4. Deep merge CLI overrides
        5. Filter unknown keys (strict rejects, lenient strips)
        6. Construct final schema object (validates + freezes)
    """

    def load(
        self,
        schema: type[T],
        *,
        files: list[tuple[str | Path, bool]] | None = None,
        env_prefix: str | None = None,
        cli_overrides: dict[str, str] | None = None,
        strict: bool = False,
        base_dir: Path | None = None,
    ) -> T:
        """Load configuration from all layers and return a frozen, validated model.

        Args:
            schema:        The Pydantic model class (e.g., AppConfig).
            files:         List of (file_path, required) tuples.
                           Files are merged in order (later files win).
            env_prefix:    Bare prefix for env var scanning (e.g., "SEVENTHPLACE").
                           If None, env vars are not scanned.
            cli_overrides: Dict of dotted-key CLI overrides.
            strict:        If True, reject unknown keys with ValidationError.
                           If False, strip unknown keys silently.
            base_dir:      Base directory for file path traversal checks.

        Returns:
            A frozen, validated instance of the schema model.

        Raises:
            ValidationError: Type or constraint validation failed, or
                             unknown keys found in strict mode.
            ConfigFileNotFoundError: A required file is missing.
            PathTraversalError: A file path escapes the base directory.
            ParseError: A file contains invalid YAML/JSON.
        """
        # Step 1: Defaults
        merged = DefaultProvider().load(schema)

        # Step 2: File overrides
        if files:
            file_provider = FileProvider()
            for file_path, required in files:
                file_data = file_provider.load(
                    file_path, required=required, base_dir=base_dir
                )
                merged = deep_merge(merged, file_data)

        # Step 3: Env overrides
        if env_prefix is not None:
            env_data = EnvProvider().load(env_prefix)
            merged = deep_merge(merged, env_data)

        # Step 4: CLI overrides
        if cli_overrides:
            cli_data = CLIProvider().load(cli_overrides)
            merged = deep_merge(merged, cli_data)

        # Step 5: Handle unknown keys
        if strict:
            unknown = _find_unknown_keys(merged, schema)
            if unknown:
                raise ValidationError(
                    f"Unknown configuration keys: {', '.join(unknown)}",
                    field_path=unknown[0],
                )
        else:
            merged = _strip_unknown_keys(merged, schema)

        # Step 6: Construct and validate
        try:
            return schema(**merged)
        except pydantic.ValidationError as exc:
            raise ValidationError(
                _format_pydantic_error(exc)
            ) from exc


def _find_unknown_keys(
    data: dict[str, Any],
    schema: type[pydantic.BaseModel],
    prefix: str = "",
) -> list[str]:
    """Recursively find keys in data that are not fields of the schema."""
    unknown: list[str] = []
    schema_fields = schema.model_fields

    for key in data:
        field_path = key if not prefix else f"{prefix}.{key}"
        if key not in schema_fields:
            unknown.append(field_path)
        elif isinstance(data[key], dict):
            field_info = schema_fields[key]
            field_type = field_info.annotation
            if (
                field_type is not None
                and isinstance(field_type, type)
                and issubclass(field_type, pydantic.BaseModel)
            ):
                unknown.extend(
                    _find_unknown_keys(data[key], field_type, field_path)
                )

    return unknown


def _strip_unknown_keys(
    data: dict[str, Any],
    schema: type[pydantic.BaseModel],
) -> dict[str, Any]:
    """Recursively remove keys from data that are not fields of the schema."""
    schema_fields = schema.model_fields
    result: dict[str, Any] = {}

    for key, value in data.items():
        if key not in schema_fields:
            continue
        if isinstance(value, dict):
            field_info = schema_fields[key]
            field_type = field_info.annotation
            if (
                field_type is not None
                and isinstance(field_type, type)
                and issubclass(field_type, pydantic.BaseModel)
            ):
                result[key] = _strip_unknown_keys(value, field_type)
            else:
                result[key] = value
        else:
            result[key] = value

    return result


def _format_pydantic_error(exc: pydantic.ValidationError) -> str:
    """Format a Pydantic ValidationError into a user-friendly message.

    Includes field path and expected type. Never includes the raw value.
    """
    parts: list[str] = []
    for error in exc.errors():
        loc = ".".join(str(segment) for segment in error["loc"])
        msg = error["msg"]
        err_type = error["type"]
        parts.append(f"{loc}: {msg} (type={err_type})")
    return "; ".join(parts)

File: typescript/src/manager.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * ConfigManager — orchestrates the layered configuration load pipeline.
 */

import { z } from "zod";
import { ValidationError } from "./errors.js";
import { deepMerge } from "./merge.js";
import { deepFreeze, getSchemaShape } from "./schema.js";
import { CLIProvider } from "./providers/cli.js";
import { DefaultProvider } from "./providers/defaults.js";
import { EnvProvider } from "./providers/env.js";
import { FileProvider } from "./providers/file.js";

export interface LoadOptions {
  /**
   * List of [filePath, required] tuples.
   * Files are merged in order (later files win).
   */
  files?: Array<[string, boolean]>;

  /** Bare prefix for env var scanning (e.g., "SEVENTHPLACE"). */
  envPrefix?: string;

  /** Dict of dotted-key CLI overrides. */
  cliOverrides?: Record<string, string>;

  /**
   * If true, reject unknown keys with ValidationError.
   * If false, strip unknown keys silently.
   */
  strict?: boolean;

  /** Base directory for file path traversal checks. */
  baseDir?: string;
}

/**
 * Orchestrates the layered configuration load pipeline.
 *
 * Load flow:
 *   1. Extract defaults from schema (parse empty object)
 *   2. Deep merge file overrides (one or more files, in order)
 *   3. Deep merge env overrides
 *   4. Deep merge CLI overrides
 *   5. Filter unknown keys (strict rejects, lenient strips)
 *   6. Validate + freeze final config object
 */
export class ConfigManager {
  load<T>(schema: z.ZodTypeAny, options: LoadOptions = {}): T {
    const {
      files,
      envPrefix,
      cliOverrides,
      strict = false,
      baseDir,
    } = options;

    // Step 1: Defaults
    const defaultProvider = new DefaultProvider();
    let merged: Record<string, unknown> = defaultProvider.load(schema);

    // Step 2: File overrides
    if (files) {
      const fileProvider = new FileProvider();
      for (const [filePath, required] of files) {
        const fileData = fileProvider.load(filePath, { required, baseDir });
        merged = deepMerge(merged, fileData);
      }
    }

    // Step 3: Env overrides
    if (envPrefix !== undefined) {
      const envProvider = new EnvProvider();
      const envData = envProvider.load(envPrefix);
      merged = deepMerge(merged, envData);
    }

    // Step 4: CLI overrides
    if (cliOverrides) {
      const cliProvider = new CLIProvider();
      const cliData = cliProvider.load(cliOverrides);
      merged = deepMerge(merged, cliData);
    }

    // Step 5: Handle unknown keys
    if (strict) {
      const unknown = findUnknownKeys(merged, schema);
      if (unknown.length > 0) {
        throw new ValidationError(
          `Unknown configuration keys: ${unknown.join(", ")}`,
          unknown[0],
        );
      }
    } else {
      merged = stripUnknownKeys(merged, schema);
    }

    // Step 6: Validate with the schema
    // We need to use the raw (pre-defaults) schema for strict parsing,
    // but since we already have defaults merged in, we just parse.
    let result: T;
    try {
      result = schema.parse(merged) as T;
    } catch (err) {
      if (err instanceof z.ZodError) {
        throw new ValidationError(formatZodError(err));
      }
      throw err;
    }

    // Deep freeze the result
    if (typeof result === "object" && result !== null) {
      deepFreeze(result as object);
    }

    return result;
  }
}

/**
 * Recursively find keys in data that are not in the schema shape.
 */
function findUnknownKeys(
  data: Record<string, unknown>,
  schema: z.ZodTypeAny,
  prefix = "",
): string[] {
  const unknown: string[] = [];
  const shape = getSchemaShape(schema);

  if (!shape) {
    return unknown;
  }

  for (const key of Object.keys(data)) {
    const fieldPath = prefix ? `${prefix}.${key}` : key;
    if (!(key in shape)) {
      unknown.push(fieldPath);
    } else if (typeof data[key] === "object" && data[key] !== null && !Array.isArray(data[key])) {
      const fieldSchema = shape[key];
      unknown.push(
        ...findUnknownKeys(data[key] as Record<string, unknown>, fieldSchema, fieldPath),
      );
    }
  }

  return unknown;
}

/**
 * Recursively remove keys from data that are not in the schema shape.
 */
function stripUnknownKeys(
  data: Record<string, unknown>,
  schema: z.ZodTypeAny,
): Record<string, unknown> {
  const shape = getSchemaShape(schema);

  if (!shape) {
    return data;
  }

  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(data)) {
    if (!(key in shape)) {
      continue;
    }
    if (typeof value === "object" && value !== null && !Array.isArray(value)) {
      result[key] = stripUnknownKeys(value as Record<string, unknown>, shape[key]);
    } else {
      result[key] = value;
    }
  }

  return result;
}

/**
 * Format a ZodError into a user-friendly message.
 *
 * Includes field path and expected type. Never includes the raw value.
 */
function formatZodError(err: z.ZodError): string {
  return err.issues
    .map((issue) => {
      const loc = issue.path.join(".");
      return `${loc}: ${issue.message} (code=${issue.code})`;
    })
    .join("; ");
}
